File: tests/swift_static/Tests/MyAirtableStaticTests/TestFieldsDualAccess.swift
// F2.10 — Verify Fields supports dual ID/name access: IDs tried first, names
// translated via nameToId. This is the parity point with the Rust
// static/rust/struct_table.rs Fields API, and unlocks the generated
// {Table}Fields.primaryKeyId / .primaryKeyName constants.

import Foundation
import Testing

@testable import MyAirtableStatic

@Suite("Fields dual ID/name access")
struct TestFieldsDualAccess {
    // Generator-style map: field name -> field ID.
    private let nameMap: [String: String] = [
        "Primary Key": "fldPrimary",
        "Count": "fldCount",
        "Active": "fldActive",
    ]

    @Test("ID lookup returns the stored value")
    func testIdLookup() {
        var f = Fields(nameToId: nameMap)
        f.setString("fldPrimary", "hello")
        #expect(f.getString("fldPrimary") == "hello")
    }

    @Test("Name lookup translates to ID and returns the same value")
    func testNameLookup() {
        var f = Fields(nameToId: nameMap)
        f.setString("fldPrimary", "hello")
        #expect(f.getString("Primary Key") == "hello")
    }

    @Test("ID is tried first when both a name and an unrelated ID collide")
    func testIdBeatsName() {
        var f = Fields(
            [
                "fldPrimary": .string("viaId")
                // (Names are never directly stored — they translate to IDs via
                // nameToId before reaching storage.)
            ],
            nameToId: nameMap
        )
        #expect(f.getString("fldPrimary") == "viaId")
        #expect(f.getString("Primary Key") == "viaId")
    }

    @Test("Set by name stores under the translated ID, not the name itself")
    func testSetByNameTranslates() {
        var f = Fields(nameToId: nameMap)
        f.setString("Primary Key", "hello")
        #expect(f.ids == ["fldPrimary"])
        #expect(f.getString("fldPrimary") == "hello")
    }

    @Test("Unknown name without mapping stores under the raw key")
    func testUnknownName() {
        var f = Fields(nameToId: nameMap)
        f.setString("Unmapped Field", "x")
        // No nameToId entry — stored as-is.
        #expect(f.getString("Unmapped Field") == "x")
    }

    @Test("Nil set removes the entry")
    func testNilRemovesEntry() {
        var f = Fields(nameToId: nameMap)
        f.setString("Count", "42")
        #expect(f.count == 1)
        f.setString("Count", nil)
        #expect(f.count == 0)
    }

    @Test("Typed getInt handles both int and double storage")
    func testGetInt() {
        var f = Fields()
        f.setInt("x", 42)
        #expect(f.getInt("x") == 42)

        f.setDouble("y", 3.14)
        #expect(f.getInt("y") == 3)
    }

    @Test("Typed getDouble handles both int and double storage")
    func testGetDouble() {
        var f = Fields()
        f.setDouble("pi", 3.14)
        #expect(f.getDouble("pi") == 3.14)

        f.setInt("answer", 42)
        #expect(f.getDouble("answer") == 42.0)
    }

    @Test("Subscript delegates to get/set")
    func testSubscript() {
        var f = Fields(nameToId: nameMap)
        f["Primary Key"] = .string("via subscript")
        #expect(f["fldPrimary"] == .string("via subscript"))
        #expect(f["Primary Key"] == .string("via subscript"))
        f["Primary Key"] = nil
        #expect(f["fldPrimary"] == nil)
    }

    @Test("Codable round-trip preserves storage but drops nameToId")
    func testCodableRoundTrip() throws {
        var original = Fields(nameToId: nameMap)
        original.setString("Primary Key", "hello")
        original.setInt("Count", 42)

        let data = try JSONEncoder().encode(original)
        let decoded = try JSONDecoder().decode(Fields.self, from: data)

        // Storage survives via field IDs.
        #expect(decoded.getString("fldPrimary") == "hello")
        #expect(decoded.getInt("fldCount") == 42)
        // nameToId is *not* serialized (it's generator metadata, not wire data).
        #expect(decoded.getString("Primary Key") == nil)
    }

    @Test("setStrings encodes a string array")
    func testSetStrings() {
        var f = Fields()
        f.setStrings("tags", ["a", "b", "c"])
        let arr = f.getArray("tags") ?? []
        #expect(arr.count == 3)
        #expect(arr[0] == .string("a"))
    }
}

// myairtable-hwqs — set() is ID-first like get() (parity with Kotlin PR #19 fix).
@Suite("Fields set ID-first")
struct TestFieldsSetIdFirst {
    @Test("Set is ID-first when a name and an unrelated ID collide")
    func setIdFirstOnCollision() {
        var fields = Fields(
            ["fldX": .string("by-id"), "fldY": .string("by-name")],
            nameToId: ["fldX": "fldY"]
        )
        fields.set("fldX", .string("updated"))
        #expect(fields.get("fldX") == .string("updated"))
        #expect(fields.get("fldY") == .string("by-name"))
    }

    @Test("A known field ID absent from storage is still never name-translated")
    func setKnownIdSparseStorage() {
        var fields = Fields(
            [:],
            nameToId: ["fldABC": "fldDEF", "Real Name": "fldABC"]
        )
        fields.set("fldABC", .int(1))
        #expect(fields.get("fldABC") == .int(1))
        #expect(fields.get("fldDEF") == nil)
    }
}

File: tests/cpp_static/test_fields_dual_access.cpp
// Fields dual ID/name access (Java TestFieldsDualAccess parity, 14 cases).
#include <catch2/catch_test_macros.hpp>

#include "fields.hpp"

using namespace myairtable;

namespace {
Fields make_fields() {
    return Fields({{"fld001", json("hello")}, {"fld002", json(42.0)}},
                  {{"Primary Key", "fld001"}, {"Count", "fld002"}});
}
} // namespace

TEST_CASE("id lookup returns the stored value", "[fields]") {
    REQUIRE(make_fields().get("fld001") == json("hello"));
}

TEST_CASE("name lookup translates to id and returns the same value", "[fields]") {
    auto fields = make_fields();
    REQUIRE(fields.get("Primary Key") == json("hello"));
    REQUIRE(fields.get("Primary Key") == fields.get("fld001"));
}

TEST_CASE("id is tried first when a name and an unrelated id collide", "[fields]") {
    // A storage key that HAPPENS to equal a mapped name: direct hit wins.
    Fields fields({{"Primary Key", json("stored-under-raw-key")}, {"fld001", json("by-id")}},
                  {{"Primary Key", "fld001"}});
    REQUIRE(fields.get("Primary Key") == json("stored-under-raw-key"));
}

TEST_CASE("set is id-first when a name and an unrelated id collide", "[fields]") {
    Fields fields({{"Primary Key", json("raw")}, {"fld001", json("by-id")}},
                  {{"Primary Key", "fld001"}});
    fields.set("Primary Key", json("updated"));
    REQUIRE(fields.get("Primary Key") == json("updated"));
    REQUIRE(fields.get("fld001") == json("by-id")); // untouched
}

TEST_CASE("set treats known id as id even when absent from storage", "[fields]") {
    Fields fields({}, {{"Primary Key", "fld001"}});
    fields.set("fld001", json("x"));
    REQUIRE(fields.get("Primary Key") == json("x"));
}

TEST_CASE("set by name stores under the translated id, not the name itself", "[fields]") {
    Fields fields({}, {{"Primary Key", "fld001"}});
    fields.set("Primary Key", json("x"));
    REQUIRE(fields.to_map().contains("fld001"));
    REQUIRE_FALSE(fields.to_map().contains("Primary Key"));
}

TEST_CASE("unknown name without mapping stores under the raw key", "[fields]") {
    Fields fields;
    fields.set("Mystery", json(1));
    REQUIRE(fields.to_map().contains("Mystery"));
}

TEST_CASE("remove drops the entry entirely", "[fields]") {
    auto fields = make_fields();
    fields.remove("Primary Key"); // via name
    REQUIRE_FALSE(fields.has("fld001"));
    REQUIRE(fields.count() == 1);
}

TEST_CASE("json null is stored as JSON null to clear the field server-side", "[fields]") {
    auto fields = make_fields();
    fields.set("Primary Key", json(nullptr));
    REQUIRE(fields.has("fld001")); // still present in the payload
    REQUIRE(fields.get("fld001").is_null());
    REQUIRE(fields.count() == 2);
}

TEST_CASE("typed getters handle both int and double storage", "[fields]") {
    Fields fields({{"fldI", json(7)}, {"fldD", json(7.5)}});
    REQUIRE(fields.get_long("fldI") == 7);
    REQUIRE(fields.get_double("fldI") == 7.0);
    REQUIRE(fields.get_long("fldD") == 7); // truncating, C# (long) cast parity
    REQUIRE(fields.get_double("fldD") == 7.5);
}

TEST_CASE("typed string getter rejects non-strings", "[fields]") {
    auto fields = make_fields();
    REQUIRE(fields.get_string("fld001") == "hello");
    REQUIRE(fields.get_string("fld002") == std::nullopt); // number, not string
    REQUIRE(fields.get_string("fldMissing") == std::nullopt);
    REQUIRE(fields.get_bool("fld001") == std::nullopt);
}

TEST_CASE("set_strings encodes a string array", "[fields]") {
    Fields fields;
    fields.set_strings("fldL", std::vector<std::string>{"rec1", "rec2"});
    REQUIRE(fields.get("fldL") == json::parse(R"(["rec1","rec2"])"));
    REQUIRE(fields.get_array("fldL")->size() == 2);
}

TEST_CASE("serialization round-trip preserves storage but drops name-to-id", "[fields]") {
    auto fields = make_fields();
    json encoded = fields;
    REQUIRE(encoded == json::parse(R"({"fld001":"hello","fld002":42.0})"));
    auto decoded = encoded.get<Fields>();
    REQUIRE(decoded == fields); // equality is storage-only
    REQUIRE(decoded.name_to_id().empty());
    REQUIRE(decoded.get("Primary Key").is_null()); // mapping gone until re-attached
}

TEST_CASE("missing fields object decodes without throwing", "[fields]") {
    REQUIRE(json(nullptr).get<Fields>().count() == 0);
    REQUIRE(json::object().get<Fields>().count() == 0);
}
